### ISSUE TITLE
Working with two inputs

### ISSUE THREAD

**Issue by wwzhou** [Dobiasd/frugally-deep#104]
I have two models trained with Keras: A and B.  A takes only images as inputs, while B takes images as well as a vector of floats as inputs.  I have successfully made Model A work from C++ through Frugally-Deep, but not Model B.

By following and modifying the examples from the web page, what I did was:

```
const auto imgin = fdeep::tensor5_from_bytes(rszimg.ptr(), rszimg.rows, rszimg.cols, rszimg.channels(), 0.0f, 1.0f);    // where rszimg is an image in OpenCV mat format.

fdeep::shape5 tensor5_shape(1, 1, vectorleng, 1, 1);
fdeep::tensor5 fvect(tensor5_shape, 0.0f);
for (int ff = 0; ff < vectorleng; ++ff) 
    fvect.set_yxz(ff, 0, 0, originalvector[ff]);

const auto result = model.predict({ imgin, fvect });

```
When stepping over the last line, I got a memory exception error.

Could you show me an example of how to do this correctly?

**Comment by Dobiasd** [comment]
Hi Will,

in general your code looks correct to me. The case of providing multiple input tensors to a model is also covered in the tests, and I've used this in the past myself in a project. So I expect this to work.

Can you post a minimal (but complete) example (with Python code) to reproduce the error? Or upload your model file (`.h5` and `.json`)?

**Comment by Dobiasd** [comment]
> Could you show me an example of how to do this correctly?

Sure, hope that helps. :)

```python3
# create_model.py

import numpy as np
from keras.layers import Input, Concatenate, Flatten, Dense
from keras.models import Model

input_1 = Input(shape=(240, 320, 3))
input_2 = Input(shape=(100, 1, 1))

output = Dense(4)(Concatenate()([Flatten()(input_1), Flatten()(input_2)]))

model = Model(inputs=[input_1, input_2], outputs=output)
model.compile(loss='categorical_crossentropy', optimizer='nadam')

model.save('keras_model.h5', include_optimizer=False)
```

```cpp
// main.cpp

#include <fdeep/fdeep.hpp>

int main()
{
    const auto model = fdeep::load_model("fdeep_model.json");
    const fdeep::tensor5 image(fdeep::shape5(1, 1, 240, 320, 3), 0.0f);
    const fdeep::tensor5 values(fdeep::shape5(1, 1, 100, 1, 1), 0.0f);
    const auto result = model.predict({image, values});
    std::cout << fdeep::show_tensor5s(result) << std::endl;
}
```

```bash
python3 create_model.py

python3 convert_model.py keras_model.h5 fdeep_model.json

g++ -std=c++14 -O3 main.cpp

./a.out
```

output:

```
Loading json ... done. elapsed time: 0.089034 s
Running test 1 of 1 ... done. elapsed time: 0.002383 s
Loading, constructing, testing of fdeep_model.json took 0.134049 s overall.
[[[[0.0000]], [[0.0000]], [[0.0000]], [[0.0000]]]]
```

**Comment by wwzhou** [comment]
Hi, Dobias:

Thanks a lot for the prompt response!

I studied your example but couldn't find anything in my code that is obviously deviant from the way you did it.  So I included the two model files below.  Hope they will help you answer my question.

By the way, the example you showed will be very helpful to new users like me.  Why don't you make it one of the examples on the documentation page?

Best,

Will


[181012-4-2-fdeep-model.zip](https://github.com/Dobiasd/frugally-deep/files/2664509/181012-4-2-fdeep-model.zip)
[181012-4-2-model-no-optimizer.zip](https://github.com/Dobiasd/frugally-deep/files/2664511/181012-4-2-model-no-optimizer.zip)

**Comment by Dobiasd** [comment]
Just tried out your mode `181012-4-2-fdeep-model.json`, and it worked here.

Code:
```cpp
// main.cpp
#include <fdeep/fdeep.hpp>
int main()
{
    const auto model = fdeep::load_model("181012-4-2-fdeep-model.json");
    const fdeep::tensor5 image(fdeep::shape5(1, 1, 128, 128, 3), 0.0f);
    const fdeep::tensor5 values(fdeep::shape5(1, 1, 1, 1, 125), 0.0f);
    const auto result = model.predict({image, values});
    std::cout << fdeep::show_tensor5s(result) << std::endl;
}
```

Commands:
```
g++ -std=c++14 -O3 main.cpp
./a.out
```

Output:
```
Loading json ... done. elapsed time: 0.286811 s
Running test 1 of 1 ... done. elapsed time: 0.059628 s
Loading, constructing, testing of 181012-4-2-fdeep-model.json took 0.492499 s overall.
[[[[0.0035]], [[0.0012]], [[0.0010]], [[0.0009]], [[0.9174]], [[0.0655]], [[0.0085]], [[0.0020]], [[0.0000]]]]
```

Could you please provide a **complete** minimal example in C++, that I can compile and run (as is) to reproduce your error?

---

> Why don't you make it one of the examples on the documentation page?

Thanks for the suggestion. I [just added](https://github.com/Dobiasd/frugally-deep/commit/e7e10d3c3d746d44d7f142c47243bf8f46d07c5f) this example to the README file. :)

**Comment by Dobiasd** [comment]
> it reported a case where the error msg was "Invalid divisor in softmax layer..." I have already set "#define FDEEP_FLOAT_TYPE double" before the "#include <fdeep/fdeep.hpp>" line. Have you seen this before?

Are you using an old version of frugally-deep? Since [this commit](https://github.com/Dobiasd/frugally-deep/commit/b10040641bd199ea6c8d1d083b58884ff4fa4313) it should not longer happen. Please update frugally-deep and try again. :)

---

> This leads to a bigger question: how much difference shall one expect between the predictions from Python vs. from Frugally-Deep? Before this, I already noticed some (a small portion of the whole) samples were classified differently by the same model on the two sides.

I've not yet seen significant differences. Maybe the ones you observe have to do with the borderline case in the softmax layer. Updating to the latest version might also fix this issue for you. Please try it out and let me know if it helped.

---

> Thanks also for adding the example. But could you use the second example above instead of the first one, because it seems to me it's important to use the shape of (1, 1, 1, 1, N) for the vector input?

This is only because of how you build your model. Both versions are equally valid. Currently my example uses the following:

```python
input_2 = Input(shape=(100, 1, 1))
```

```cpp
const fdeep::tensor5 values(fdeep::shape5(1, 1, 100, 1, 1), 0.0f);
```

Your model however seems to be like this:

```python
input_2 = Input(shape=(1, 1, 100))
```

```cpp
const fdeep::tensor5 values(fdeep::shape5(1, 1, 1, 1, 100), 0.0f);
```

I don't see one approach being technically superior to the other. The only reason to strictly prefer one, might be convention.

**Comment by wwzhou** [comment]
The new version works.  Thanks a lot, Dobias!

**Comment by Dobiasd** [comment]
:+1:
